remove cloudmaster's key from root's authorized keys as soon as the V…

aranhoide opened the pull request:
…PS is set up

To make this possible, some code has been changed not to use that key any more.

I'm convinced this vulnerability is currently highly theoretical.  To obtain this key, an attacker probably needs to have gained root access to a cloudmaster or login access to our Vultr or DO admin page, both of which are about as dangerous as this key.

It's still wise to remove it, lest it becomes more of a vulnerability in the future.

I've tested this in my test cloudmaster.

oxtoacart commented on salt/security.sls:
```diff
@@ -39,3 +39,10 @@ reload-sshd-on-password-disable:
         - file: disable-password-authentication
         - file: disable-challenge-response-authentication
+
+/root/.ssh/authorized_keys:
```
What is this doing?  Might warrant a comment, since it's non-obvious.

aranhoide replied:
This removes the `cloudmaster` public key from the root's authorized keys.  That key is identified by having the email lanterncyborg@gmail.com.

I could delete authorized_keys altogether, but if we ever want to accept any SSH key for the root user and we forget about this, it could get really frustrating to debug.

I'll add a comment to this effect when I'm back at my computer.

aranhoide replied:
The rationale is that by the time we have successfully installed Salt and set up this minion with its cloudmaster, the ssh key has performed its function and we don't need it anymore.  The rest of this PR is about making that assumption stand.

oxtoacart commented on lib/do_util.py:
```diff
@@ -12,7 +12,6 @@
 
 
 ssh_tmpl = "ssh -o StrictHostKeyChecking=no -i /etc/salt/cloudmaster.id_rsa root@%s '%%s'"
```
This still references root, but ssh_tmpl doesn't appear to be used anywhere so you could probably just get rid of it.

oxtoacart replied:
Oops, strike that, it is still used.

oxtoacart replied:
Actually, double reversal - there's an ssh_tmpl function that's used, but the ssh_tmpl string appears to be unused.

aranhoide replied:
Yes, the cloudmaster key is used for initializing the VPSs, and only once Salt is set up and connected to the master is the key deleted.

In `vultr_util` we need to refer to the key in code because we initialize the VPS directly using the Vultr API.  For DO this initialization is performed by salt-cloud, so while the key is [still used](https://github.com/getlantern/lantern_aws/blob/master/salt/salt_cloud/cloud.providers#L6) we don't need to mention it here.

I'll delete this line, thanks!

oxtoacart commented:
@aranhoide I assume you tested this.  Other than adding a comment about the authorized_keys removal and maybe getting rid of the ssh_tmpl string, this LGTM.

aranhoide commented:
@oxtoacart : I'd tested this up until server creation, which involves checkfallbacks check.  Alas, this leaves one last step untested, and I've just realized there's a problem there.  There are some spurious `!!python/unicode` YAML extension markers.  I'll fix this.

aranhoide commented on lib/vps_util.py:
```diff
@@ -67,36 +69,46 @@ def highstate_pid(name):
-                        print("Fallback check failed; retrying...")
-                    else:
-                        print("VPS up!")
-                        return yaml.load(access_data)
```
Actually, I've just remembered that there was a good reason why I was parsing this with the YAML library and not the JSON one.  Turns out that JSON returns all strings as unicode objects, which produce `!!python/unicode` markers when encoding to YAML.  I'll revert this particular change and add a comment to that effect.

aranhoide replied:
Actually, I'll do a more proper fix, using `yaml.safe_dump` when reencoding the file.

aranhoide commented:
@oxtoacart: I've added the comment you suggest and got rid of ssh_tmpl.  I've also fixed a bug (`!!python/unicode` markers creeping into the config uploaded to redis) which I had missed because it happens after the checkfallbacks check.  I've tested this final change (and this time around I checked the config in redis).

Still looks good to you?

oxtoacart commented on lib/vps_util.py:
```diff
@@ -67,36 +69,46 @@ def highstate_pid(name):
+                try:
+                    adstr = subprocess.check_output(['salt', name, 'cmd.run', 'cat /home/lantern/access_data.json'])
+                except subprocess.CalledProcessError:
+                    continue
```
Might be worth logging the error text.

aranhoide replied:
Yes, will do.

oxtoacart commented:
You might want to log the subprocess exception when we can't cat access_data.json, just in case there's something useful in there for debugging.  Otherwise, LGTM and you can pull whenever you like.

aranhoide commented:
@oxtoacart: thanks for your review!

aranhoide commented:
Whoops, I merged without pushing the printout of the subprocess exception.  I'll just push that to master.
